File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.rs
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn good_nodes(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let Some(root) = root else {
            return 0;
        };

        let mut stack = vec![(root, i32::MIN)];
        let mut answer = 0;
        while let Some((node, path_max)) = stack.pop() {
            let (value, left, right) = {
                let borrowed = node.borrow();
                (borrowed.val, borrowed.left.clone(), borrowed.right.clone())
            };
            if value >= path_max {
                answer += 1;
            }
            let current_max = value.max(path_max);
            if let Some(right) = right {
                stack.push((right, current_max));
            }
            if let Some(left) = left {
                stack.push((left, current_max));
            }
        }
        answer
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.py
class Solution:
    def maxPathSum(self, root: TreeNode | None) -> int:
        if root is None:
            return 0

        stack = [[root, 0, 0]]
        last_gain = 0
        best = root.val

        while stack:
            node, state, _ = stack[-1]
            if state == 0:
                stack[-1][1] = 1
                if node.left is not None:
                    stack.append([node.left, 0, 0])
            elif state == 1:
                stack[-1][2] = last_gain if node.left is not None else 0
                stack[-1][1] = 2
                if node.right is not None:
                    stack.append([node.right, 0, 0])
            else:
                _, _, left_gain = stack.pop()
                right_gain = last_gain if node.right is not None else 0
                left_arm = max(0, left_gain)
                right_arm = max(0, right_gain)
                through = node.val + left_arm + right_arm
                best = max(best, through)
                last_gain = node.val + max(left_arm, right_arm)

        return best